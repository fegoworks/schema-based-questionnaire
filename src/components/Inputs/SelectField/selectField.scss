.SelectField {
  margin-bottom: 0.6rem;
  display: flex;
  flex-direction: column;
  width: 15rem;
}

.SelectField select {
  padding: 0.6em 0.6em;
  font-size: 1em;
  font-weight: 500;
  transition: border-color 0.25s;
}

.SelectField option {
  font-weight: 500;
}

.SelectField .error {
  color: #d44;
  margin: 0;
}

.SelectField select:focus-visible,
.SelectField select:focus {
  border-radius: 3px;
  outline: 4px auto -webkit-focus-ring-color;
  outline-color: #ddd;
}

