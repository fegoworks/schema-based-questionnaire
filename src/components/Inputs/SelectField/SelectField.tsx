import "./selectField.scss";
import { FunctionComponent } from "react";
import { FieldProps } from "../../formField";

const SelectField: FunctionComponent<FieldProps> = ({
  error,
  name,
  onBlur,
  onChange,
  touched,
  value,
  options,
  disabled,
}) => {
  return (
    <div className="SelectField">
      <select
        id={name}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        value={value}
        disabled={disabled}>
        <option value="">Select</option>
        {options?.map((option) => (
          <option key={option.value} value={option.value}>
            {option.label}
          </option>
        ))}
      </select>
      {error && touched && <p className="error">{error}</p>}
    </div>
  );
};

export default SelectField;
