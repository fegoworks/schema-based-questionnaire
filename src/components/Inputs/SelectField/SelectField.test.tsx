import { fireEvent, render } from "@testing-library/react";
import { describe, expect, vi } from "vitest";
import SelectField from "./SelectField";

const opt = [
  { label: "Onions", value: "onions" },
  { label: "Veggies", value: "veggies" },
  { label: "Carrots", value: "carrots" },
];
const defaultProps = {
  name: "types",
  options: opt,
  disabled: false,
};

describe("Component - SelectField", () => {
  let onChange: () => void;

  beforeEach(() => {
    onChange = vi.fn();
  });

  it("should render SelectField component", () => {
    const { container } = render(<SelectField {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should render a SelectField with a dropdown of options", () => {
    const { container } = render(
      <SelectField {...defaultProps} onChange={() => onChange} />
    );
    const select = container.getElementsByTagName("select")[0];
    const options = select.getElementsByTagName("option");
    // 4 options including "Select"
    expect(options[0]).toBeInTheDocument();
    expect(options[1]).toBeInTheDocument();
    expect(options[2]).toBeInTheDocument();
    expect(options[3]).toBeInTheDocument();
    expect(options).toHaveLength(4);
  });

  it("should select the correct option from dropdown", () => {
    const { container } = render(<SelectField {...defaultProps} />);
    const select = container.getElementsByTagName("select")[0];
    fireEvent.change(select, {
      target: { value: "carrots" },
    });
    let options = select.getElementsByTagName("option");
    // 4 options including "Select"
    expect(options[0].selected).toBeFalsy();
    expect(options[1].selected).toBeFalsy();
    expect(options[2].selected).toBeFalsy();
    // selected option => carrots
    expect(options[3].selected).toBeTruthy();
  });

  it("should show an error message if validation fails", () => {
    const { getByText } = render(
      <SelectField {...defaultProps} error={"Some error"} touched={true} />
    );
    expect(getByText("Some error")).toBeTruthy();
  });
});
