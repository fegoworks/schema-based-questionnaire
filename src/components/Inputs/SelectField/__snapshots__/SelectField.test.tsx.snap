// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component - SelectField > should render SelectField component 1`] = `
<div>
  <div
    class="SelectField"
  >
    <select
      id="types"
      name="types"
    >
      <option
        value=""
      >
        Select
      </option>
      <option
        value="onions"
      >
        Onions
      </option>
      <option
        value="veggies"
      >
        Veggies
      </option>
      <option
        value="carrots"
      >
        Carrots
      </option>
    </select>
  </div>
</div>
`;
